{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from os.path import exists\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import LabelEncoder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(285, 46)\n",
      "  !Sample_geo_accession           !Sample_status !Sample_submission_date  \\\n",
      "0          \"GSM1662534\"  \"Public on Oct 30 2015\"           \"Apr 21 2015\"   \n",
      "1          \"GSM1662535\"  \"Public on Oct 30 2015\"           \"Apr 21 2015\"   \n",
      "2          \"GSM1662536\"  \"Public on Oct 30 2015\"           \"Apr 21 2015\"   \n",
      "3          \"GSM1662537\"  \"Public on Oct 30 2015\"           \"Apr 21 2015\"   \n",
      "4          \"GSM1662538\"  \"Public on Oct 30 2015\"           \"Apr 21 2015\"   \n",
      "\n",
      "  !Sample_last_update_date !Sample_type !Sample_channel_count  \\\n",
      "0            \"May 15 2019\"        \"SRA\"                   \"1\"   \n",
      "1            \"May 15 2019\"        \"SRA\"                   \"1\"   \n",
      "2            \"May 15 2019\"        \"SRA\"                   \"1\"   \n",
      "3            \"May 15 2019\"        \"SRA\"                   \"1\"   \n",
      "4            \"May 15 2019\"        \"SRA\"                   \"1\"   \n",
      "\n",
      "  !Sample_source_name_ch1 !Sample_organism_ch1 !Sample_characteristics_ch1  \\\n",
      "0    \"3-Breast-Her2-ampl\"       \"Homo sapiens\"             \"tissue: blood\"   \n",
      "1           \"8-Breast-WT\"       \"Homo sapiens\"             \"tissue: blood\"   \n",
      "2   \"10-Breast-Her2-ampl\"       \"Homo sapiens\"             \"tissue: blood\"   \n",
      "3            \"Breast-100\"       \"Homo sapiens\"             \"tissue: blood\"   \n",
      "4   \"15-Breast-Her2-ampl\"       \"Homo sapiens\"             \"tissue: blood\"   \n",
      "\n",
      "  !Sample_characteristics_ch1.1  ... !Sample_instrument_model  \\\n",
      "0     \"cell type: Thrombocytes\"  ...    \"Illumina HiSeq 2500\"   \n",
      "1     \"cell type: Thrombocytes\"  ...    \"Illumina HiSeq 2500\"   \n",
      "2     \"cell type: Thrombocytes\"  ...    \"Illumina HiSeq 2500\"   \n",
      "3     \"cell type: Thrombocytes\"  ...    \"Illumina HiSeq 2500\"   \n",
      "4     \"cell type: Thrombocytes\"  ...    \"Illumina HiSeq 2500\"   \n",
      "\n",
      "  !Sample_library_selection !Sample_library_source !Sample_library_strategy  \\\n",
      "0                    \"cDNA\"       \"transcriptomic\"                \"RNA-Seq\"   \n",
      "1                    \"cDNA\"       \"transcriptomic\"                \"RNA-Seq\"   \n",
      "2                    \"cDNA\"       \"transcriptomic\"                \"RNA-Seq\"   \n",
      "3                    \"cDNA\"       \"transcriptomic\"                \"RNA-Seq\"   \n",
      "4                    \"cDNA\"       \"transcriptomic\"                \"RNA-Seq\"   \n",
      "\n",
      "                                    !Sample_relation  \\\n",
      "0  \"BioSample: https://www.ncbi.nlm.nih.gov/biosa...   \n",
      "1  \"BioSample: https://www.ncbi.nlm.nih.gov/biosa...   \n",
      "2  \"BioSample: https://www.ncbi.nlm.nih.gov/biosa...   \n",
      "3  \"BioSample: https://www.ncbi.nlm.nih.gov/biosa...   \n",
      "4  \"BioSample: https://www.ncbi.nlm.nih.gov/biosa...   \n",
      "\n",
      "                                  !Sample_relation.1  \\\n",
      "0  \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SR...   \n",
      "1  \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SR...   \n",
      "2  \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SR...   \n",
      "3  \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SR...   \n",
      "4  \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SR...   \n",
      "\n",
      "  !Sample_supplementary_file_1 !series_matrix_table_begin      \"ID_REF\"  \\\n",
      "0                       \"NONE\"                        NaN  \"GSM1662534\"   \n",
      "1                       \"NONE\"                        NaN  \"GSM1662535\"   \n",
      "2                       \"NONE\"                        NaN  \"GSM1662536\"   \n",
      "3                       \"NONE\"                        NaN  \"GSM1662537\"   \n",
      "4                       \"NONE\"                        NaN  \"GSM1662538\"   \n",
      "\n",
      "  !series_matrix_table_end  \n",
      "0                      NaN  \n",
      "1                      NaN  \n",
      "2                      NaN  \n",
      "3                      NaN  \n",
      "4                      NaN  \n",
      "\n",
      "[5 rows x 46 columns]\n"
     ]
    }
   ],
   "source": [
    "series_matrix = pd.read_csv('data/GSE68086_series_matrix.csv')\n",
    "print(series_matrix.shape)\n",
    "print(series_matrix.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57736, 286)\n",
      "        Unnamed: 0  3-Breast-Her2-ampl  8-Breast-WT  10-Breast-Her2-ampl  \\\n",
      "0  ENSG00000000003                   0            0                    0   \n",
      "1  ENSG00000000005                   0            0                    0   \n",
      "2  ENSG00000000419                  44           14                   16   \n",
      "3  ENSG00000000457                  26            1                   14   \n",
      "4  ENSG00000000460                  81           98                   18   \n",
      "\n",
      "   Breast-100  15-Breast-Her2-ampl  16-Breast-WT  21-Breast-WT  \\\n",
      "0           0                   17             0             0   \n",
      "1           0                    0             0             0   \n",
      "2           8                    9             0           139   \n",
      "3           0                    4            20             1   \n",
      "4          17                    0            20           144   \n",
      "\n",
      "   33-Breast-Her2-ampl  42-Breast-Her2-ampl  ...  MGH-NSCLC-L20-TR500  \\\n",
      "0                    0                    0  ...                    0   \n",
      "1                    0                    0  ...                    0   \n",
      "2                  108                   55  ...                    2   \n",
      "3                   26                   24  ...                    0   \n",
      "4                   26                   97  ...                   23   \n",
      "\n",
      "   MGH-NSCLC-L22-TR462  MGH-NSCLC-L23-TR524  MGH-NSCLC-L25-TR480  \\\n",
      "0                    5                   10                    9   \n",
      "1                    0                    0                    0   \n",
      "2                   41                   40                   15   \n",
      "3                    0                    0                    0   \n",
      "4                    1                    6                   22   \n",
      "\n",
      "   MGH-NSCLC-L39-TR519  MGH-NSCLC-L40-TR520  MGH-NSCLC-L51-TR521  \\\n",
      "0                    0                    0                    0   \n",
      "1                    0                    0                    0   \n",
      "2                    1                    0                   13   \n",
      "3                    0                    0                    0   \n",
      "4                   30                    0                    1   \n",
      "\n",
      "   MGH-NSCLC-L58-TR525  MGH-NSCLC-L59-TR522  MGH-NSCLC-L65-TR523  \n",
      "0                    0                    0                    0  \n",
      "1                    0                    0                    0  \n",
      "2                    0                    0                   10  \n",
      "3                    0                    0                    0  \n",
      "4                   50                    6                    2  \n",
      "\n",
      "[5 rows x 286 columns]\n"
     ]
    }
   ],
   "source": [
    "TEP_data = pd.read_csv('data/GSE68086_TEP_data_matrix.csv')\n",
    "print(TEP_data.shape)\n",
    "print(TEP_data.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                Sample  Cancer_Type_Breast  Cancer_Type_CRC  Cancer_Type_GBM  \\\n",
      "0   3-Breast-Her2-ampl                True            False            False   \n",
      "1          8-Breast-WT                True            False            False   \n",
      "2  10-Breast-Her2-ampl                True            False            False   \n",
      "3           Breast-100                True            False            False   \n",
      "4  15-Breast-Her2-ampl                True            False            False   \n",
      "\n",
      "   Cancer_Type_HC  Cancer_Type_Hepatobiliary  Cancer_Type_Lung  \\\n",
      "0           False                      False             False   \n",
      "1           False                      False             False   \n",
      "2           False                      False             False   \n",
      "3           False                      False             False   \n",
      "4           False                      False             False   \n",
      "\n",
      "   Cancer_Type_Pancreas  Cancer_Type_mutational subclass: EGFR  \\\n",
      "0                 False                                  False   \n",
      "1                 False                                  False   \n",
      "2                 False                                  False   \n",
      "3                 False                                  False   \n",
      "4                 False                                  False   \n",
      "\n",
      "   Cancer_Type_mutational subclass: HER2+  \\\n",
      "0                                   False   \n",
      "1                                   False   \n",
      "2                                   False   \n",
      "3                                   False   \n",
      "4                                   False   \n",
      "\n",
      "   Cancer_Type_mutational subclass: KRAS  \\\n",
      "0                                  False   \n",
      "1                                  False   \n",
      "2                                  False   \n",
      "3                                  False   \n",
      "4                                  False   \n",
      "\n",
      "   Cancer_Type_mutational subclass: PIK3CA  \\\n",
      "0                                    False   \n",
      "1                                    False   \n",
      "2                                    False   \n",
      "3                                    False   \n",
      "4                                    False   \n",
      "\n",
      "   Cancer_Type_mutational subclass: wt  \n",
      "0                                False  \n",
      "1                                False  \n",
      "2                                False  \n",
      "3                                False  \n",
      "4                                False  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/x1/4ffcbyys3j9b4z3p5kdqn3jh0000gn/T/ipykernel_83460/698329259.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  series_matrix_source_cancer['Sample'] = series_matrix_source_cancer['Sample'].str.replace('\"', '')\n",
      "/var/folders/x1/4ffcbyys3j9b4z3p5kdqn3jh0000gn/T/ipykernel_83460/698329259.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  series_matrix_source_cancer['Cancer_Type'] = series_matrix_source_cancer['Cancer_Type'].str.replace('cancer type: ', '')\n",
      "/var/folders/x1/4ffcbyys3j9b4z3p5kdqn3jh0000gn/T/ipykernel_83460/698329259.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  series_matrix_source_cancer['Cancer_Type'] = series_matrix_source_cancer['Cancer_Type'].str.replace('\"', '')\n"
     ]
    }
   ],
   "source": [
    "# Keep only the sample source and cancer type columns\n",
    "series_matrix_source_cancer = series_matrix[['!Sample_source_name_ch1', '!Sample_characteristics_ch1.3']]\n",
    "# Rename the columns\n",
    "series_matrix_source_cancer.columns = ['Sample', 'Cancer_Type']\n",
    "# Delete the double quotes in the source\n",
    "series_matrix_source_cancer['Sample'] = series_matrix_source_cancer['Sample'].str.replace('\"', '')\n",
    "# Delete the cancer type in str\n",
    "series_matrix_source_cancer['Cancer_Type'] = series_matrix_source_cancer['Cancer_Type'].str.replace('cancer type: ', '')\n",
    "series_matrix_source_cancer['Cancer_Type'] = series_matrix_source_cancer['Cancer_Type'].str.replace('\"', '')\n",
    "# One hot encode the cancer type\n",
    "series_matrix_source_cancer = pd.get_dummies(series_matrix_source_cancer, columns=['Cancer_Type'])\n",
    "print(series_matrix_source_cancer.head())\n",
    "\n",
    "# Write the data to a csv file\n",
    "series_matrix_source_cancer.to_csv('data/series_matrix_source_cancer.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           gene_ID              Sample  FPKM  Cancer_Type_Breast  \\\n",
      "0  ENSG00000000003  3-Breast-Her2-ampl     0                True   \n",
      "1  ENSG00000000005  3-Breast-Her2-ampl     0                True   \n",
      "2  ENSG00000000419  3-Breast-Her2-ampl    44                True   \n",
      "3  ENSG00000000457  3-Breast-Her2-ampl    26                True   \n",
      "4  ENSG00000000460  3-Breast-Her2-ampl    81                True   \n",
      "\n",
      "   Cancer_Type_CRC  Cancer_Type_GBM  Cancer_Type_HC  \\\n",
      "0            False            False           False   \n",
      "1            False            False           False   \n",
      "2            False            False           False   \n",
      "3            False            False           False   \n",
      "4            False            False           False   \n",
      "\n",
      "   Cancer_Type_Hepatobiliary  Cancer_Type_Lung  Cancer_Type_Pancreas  \\\n",
      "0                      False             False                 False   \n",
      "1                      False             False                 False   \n",
      "2                      False             False                 False   \n",
      "3                      False             False                 False   \n",
      "4                      False             False                 False   \n",
      "\n",
      "   Cancer_Type_mutational subclass: EGFR  \\\n",
      "0                                  False   \n",
      "1                                  False   \n",
      "2                                  False   \n",
      "3                                  False   \n",
      "4                                  False   \n",
      "\n",
      "   Cancer_Type_mutational subclass: HER2+  \\\n",
      "0                                   False   \n",
      "1                                   False   \n",
      "2                                   False   \n",
      "3                                   False   \n",
      "4                                   False   \n",
      "\n",
      "   Cancer_Type_mutational subclass: KRAS  \\\n",
      "0                                  False   \n",
      "1                                  False   \n",
      "2                                  False   \n",
      "3                                  False   \n",
      "4                                  False   \n",
      "\n",
      "   Cancer_Type_mutational subclass: PIK3CA  \\\n",
      "0                                    False   \n",
      "1                                    False   \n",
      "2                                    False   \n",
      "3                                    False   \n",
      "4                                    False   \n",
      "\n",
      "   Cancer_Type_mutational subclass: wt  \n",
      "0                                False  \n",
      "1                                False  \n",
      "2                                False  \n",
      "3                                False  \n",
      "4                                False  \n"
     ]
    }
   ],
   "source": [
    "# Load the data\n",
    "patient_data = pd.read_csv('data/series_matrix_source_cancer.csv')\n",
    "rna_expression = pd.read_csv('data/GSE68086_TEP_data_matrix.csv', index_col=0)\n",
    "# print(patient_data.head())\n",
    "\n",
    "# Reset the index to make the index a regular column\n",
    "rna_expression.reset_index(inplace=True)\n",
    "# Rename the index column to 'gene_ID'\n",
    "rna_expression.rename(columns={'index': 'gene_ID'}, inplace=True)\n",
    "\n",
    "# Reshape data from wide to long format\n",
    "rna_expression_long = rna_expression.melt(id_vars=['gene_ID'], var_name='Sample', value_name='FPKM')\n",
    "\n",
    "# print(rna_expression_long.head())\n",
    "\n",
    "# Merge the patient data with the rna expression data\n",
    "patient_data['Sample'] = patient_data['Sample']\n",
    "merged_data = pd.merge(rna_expression_long, patient_data, on='Sample')\n",
    "print(merged_data.head())\n",
    "\n",
    "# Write the data to a csv file\n",
    "merged_data.to_csv('data/merged_data.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualize Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = pd.read_csv('data/merged_data.csv')\n",
    "# Compare the difference in gene expression between breast cancer and normal samples\n",
    "# Select the breast cancer and normal samples\n",
    "breast_cancer_patients = merged_data[(merged_data['Cancer_Type_Breast'] == 1)]\n",
    "healthy_controls = merged_data[(merged_data['Cancer_Type_Normal'] == 1)]\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
